### ISSUE TITLE
loss going to nan from step 1

### ISSUE THREAD

**Issue by 1NNcoder** [thtrieu/darkflow#151]
I am training on a 17 classes and have ~6000 training samples. 
I set my `filters=110` for last conv layer as stated.

The cmd I am using this
`./flow --train --dataset /home/ubuntu/datasets/signDatabasePublicFramesOnly/img/ --annotation /home/ubuntu/datasets/signDatabasePublicFramesOnly/anno/ --model cfg/yolo-17c.cfg --load bin/yolo.weights --keep 5 --epoch 30000 --save 1000 --lr 0.00002  --batch 16 --gpu .9`

My loss is going to nan right from step 1
Any ideas how to get around this?
> I tensorflow/core/common_runtime/gpu/gpu_device.cc:906] DMA: 0
> I tensorflow/core/common_runtime/gpu/gpu_device.cc:916] 0:   Y
> I tensorflow/core/common_runtime/gpu/gpu_device.cc:975] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Graphics Device, pci bus id: 0000:03:00.0)
> Finished in 7.160877466201782s
> 
> Enter training ...
> Dataset of 5904 instance(s)
> /home/ubuntu/darkflow/net/yolov2/data.py:41: RuntimeWarning: invalid value encountered in sqrt
>   obj[4] = np.sqrt(obj[4])
> Training statistics:
>         Learning rate : 2e-05
>         Batch size    : 16
>         Epoch number  : 30000
>         Backup every  : 1000
> step 1 - loss nan - moving ave loss nan
> step 2 - loss nan - moving ave loss nan
> step 3 - loss nan - moving ave loss nan
> step 4 - loss nan - moving ave loss nan
> I tensorflow/core/common_runtime/gpu/pool_allocator.cc:247] PoolAllocator: After 2553 get requests, put_count=2433 evicted_count=1000 eviction_rate=0.411015 and unsatisfied allocation rate=0.477869
> I tensorflow/core/common_runtime/gpu/pool_allocator.cc:259] Raising pool_size_limit_ from 100 to 110
> step 5 - loss nan - moving ave loss nan
> step 6 - loss nan - moving ave loss nan
> step 7 - loss nan - moving ave loss nan
> step 8 - loss nan - moving ave loss nan
> step 9 - loss nan - moving ave loss nan
> step 10 - loss nan - moving ave loss nan
> step 11 - loss nan - moving ave loss nan
> step 12 - loss nan - moving ave loss nan

@thtrieu , @Dhruv-Mohan , @abagshaw, @dontfollowmeimcrazy @imian00

Any help is appreciated. Thank you

**Comment by 1NNcoder** [comment]
Does the format of images Matter ?
I'm using jpg as input format

**Comment by imian00** [comment]
sorry i'm a beginner in this field.. but i think it would be better first you overfit your model with very small datasets and see if your model converges and then try it again. if it does not converge even if you used small datasets, then the model may have some bugs
#80

**Comment by 1NNcoder** [comment]
here is log:

> ~/darkflow  [master]$ ./flow --train --dataset /home/ubuntu/datasets/over_fit/img/ --annotation /home/ubuntu/datasets/over_fit/anno/ --model cfg/yolo-2c.cfg --load bin/yolo.weights --keep 5 --epoch 30000 --save 1000 --lr 0.00001  --batch 24 --gpu .95 --summary ~/summary --trainer adam
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcublas.so.8.0 locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcudnn.so.5 locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcufft.so.8.0 locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcuda.so.1 locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcurand.so.8.0 locally
Parsing ./cfg/yolo.cfg
Parsing cfg/yolo-2c.cfg
Loading bin/yolo.weights ...
Successfully identified 269862452 bytes
Finished in 0.010815858840942383s

Building net ...
Source | Train? | Layer description                | Output size
-------+--------+----------------------------------+---------------
       |        | input                            | (?, 416, 416, 3)
 Load  |  Yep!  | conv 3x3p1_1  +bnorm  leaky      | (?, 416, 416, 32)
 Load  |  Yep!  | maxp 2x2p0_2                     | (?, 208, 208, 32)
 Load  |  Yep!  | conv 3x3p1_1  +bnorm  leaky      | (?, 208, 208, 64)
 Load  |  Yep!  | maxp 2x2p0_2                     | (?, 104, 104, 64)
 Load  |  Yep!  | conv 3x3p1_1  +bnorm  leaky      | (?, 104, 104, 128)
 Load  |  Yep!  | conv 1x1p0_1  +bnorm  leaky      | (?, 104, 104, 64)
 Load  |  Yep!  | conv 3x3p1_1  +bnorm  leaky      | (?, 104, 104, 128)
 Load  |  Yep!  | maxp 2x2p0_2                     | (?, 52, 52, 128)
 Load  |  Yep!  | conv 3x3p1_1  +bnorm  leaky      | (?, 52, 52, 256)
 Load  |  Yep!  | conv 1x1p0_1  +bnorm  leaky      | (?, 52, 52, 128)
 Load  |  Yep!  | conv 3x3p1_1  +bnorm  leaky      | (?, 52, 52, 256)
 Load  |  Yep!  | maxp 2x2p0_2                     | (?, 26, 26, 256)
 Load  |  Yep!  | conv 3x3p1_1  +bnorm  leaky      | (?, 26, 26, 512)
 Load  |  Yep!  | conv 1x1p0_1  +bnorm  leaky      | (?, 26, 26, 256)
 Load  |  Yep!  | conv 3x3p1_1  +bnorm  leaky      | (?, 26, 26, 512)
 Load  |  Yep!  | conv 1x1p0_1  +bnorm  leaky      | (?, 26, 26, 256)
 Load  |  Yep!  | conv 3x3p1_1  +bnorm  leaky      | (?, 26, 26, 512)
 Load  |  Yep!  | maxp 2x2p0_2                     | (?, 13, 13, 512)
 Load  |  Yep!  | conv 3x3p1_1  +bnorm  leaky      | (?, 13, 13, 1024)
 Load  |  Yep!  | conv 1x1p0_1  +bnorm  leaky      | (?, 13, 13, 512)
 Load  |  Yep!  | conv 3x3p1_1  +bnorm  leaky      | (?, 13, 13, 1024)
 Load  |  Yep!  | conv 1x1p0_1  +bnorm  leaky      | (?, 13, 13, 512)
 Load  |  Yep!  | conv 3x3p1_1  +bnorm  leaky      | (?, 13, 13, 1024)
 Load  |  Yep!  | conv 3x3p1_1  +bnorm  leaky      | (?, 13, 13, 1024)
 Load  |  Yep!  | conv 3x3p1_1  +bnorm  leaky      | (?, 13, 13, 1024)
 Load  |  Yep!  | concat [16]                      | (?, 26, 26, 512)
 Load  |  Yep!  | local flatten 2x2                | (?, 13, 13, 2048)
 Load  |  Yep!  | concat [26, 24]                  | (?, 13, 13, 3072)
 Load  |  Yep!  | conv 3x3p1_1  +bnorm  leaky      | (?, 13, 13, 1024)
 Init  |  Yep!  | conv 1x1p0_1    linear           | (?, 13, 13, 35)
-------+--------+----------------------------------+---------------
GPU mode with 0.95 usage
cfg/yolo-2c.cfg loss hyper-parameters:
	H       = 13
	W       = 13
	box     = 5
	classes = 2
	scales  = [1.0, 5.0, 1.0, 1.0]
Building cfg/yolo-2c.cfg loss
Building cfg/yolo-2c.cfg train op
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE3 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.1 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX2 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use FMA instructions, but these are available on your machine and could speed up CPU computations.
I tensorflow/core/common_runtime/gpu/gpu_device.cc:885] Found device 0 with properties:
name: Graphics Device
major: 6 minor: 1 memoryClockRate (GHz) 1.582
pciBusID 0000:03:00.0
Total memory: 10.91GiB
Free memory: 10.76GiB
I tensorflow/core/common_runtime/gpu/gpu_device.cc:906] DMA: 0
I tensorflow/core/common_runtime/gpu/gpu_device.cc:916] 0:   Y
I tensorflow/core/common_runtime/gpu/gpu_device.cc:975] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Graphics Device, pci bus id: 0000:03:00.0)
Finished in 7.062250375747681s

Enter training ...

cfg/yolo-2c.cfg parsing /home/ubuntu/datasets/over_fit/anno/
Parsing for ['signalAhead', 'pedestrianCrossing']
[====================>]100%  pedestrianCrossing_1324867122.avi_image0.xml
Statistics:
signalAhead: 10
pedestrianCrossing: 12
Dataset size: 22
Result saved to net/yolo/yolo-2c.parsed
Dataset of 22 instance(s)
/home/ubuntu/darkflow/net/yolov2/data.py:40: RuntimeWarning: invalid value encountered in sqrt
  obj[3] = np.sqrt(obj[3])
/home/ubuntu/darkflow/net/yolov2/data.py:41: RuntimeWarning: invalid value encountered in sqrt
  obj[4] = np.sqrt(obj[4])
Training statistics:
	Learning rate : 1e-05
	Batch size    : 22
	Epoch number  : 30000
	Backup every  : 1000
step 1 - loss nan - moving ave loss nan
Finish 1 epoch(es)
step 2 - loss nan - moving ave loss nan
Finish 2 epoch(es)
step 3 - loss nan - moving ave loss nan
Finish 3 epoch(es)
step 4 - loss nan - moving ave loss nan
Finish 4 epoch(es)
I tensorflow/core/common_runtime/gpu/pool_allocator.cc:247] PoolAllocator: After 2553 get requests, put_count=2433 evicted_count=1000 eviction_rate=0.411015 and unsatisfied allocation rate=0.477869
I tensorflow/core/common_runtime/gpu/pool_allocator.cc:259] Raising pool_size_limit_ from 100 to 110
step 5 - loss nan - moving ave loss nan
Finish 5 epoch(es)
step 6 - loss nan - moving ave loss nan
Finish 6 epoch(es)
step 7 - loss nan - moving ave loss nan
Finish 7 epoch(es)
step 8 - loss nan - moving ave loss nan
Finish 8 epoch(es)
step 9 - loss nan - moving ave loss nan
Finish 9 epoch(es)

**Comment by 1NNcoder** [comment]
What concerns me is this from the above log
`/home/ubuntu/darkflow/net/yolov2/data.py:40: RuntimeWarning: invalid value encountered in sqrt
obj[3] = np.sqrt(obj[3])
/home/ubuntu/darkflow/net/yolov2/data.py:41: RuntimeWarning: invalid value encountered in sqrt`

**Comment by thtrieu** [comment]
`obj[3]` is supposed to be `xmax - xmin`. I think the parser got it wrong (or the annotation is wrong?), that xmin and xmax are mistaken for each other.

I do have to look into these values, or the `xml` file to see why.

**Comment by 1NNcoder** [comment]
> <?xml version="1.0" ?>
<annotation verified="yes">
  <folder>/home/ubuntu/datasets/img/</folder>
  <filename>pedestrian_1323820836.avi_image13.jpg</filename>
  <path>/home/ubuntu/datasets/img/pedestrian_1323820836.avi_image13.jpg</path>
  <source>
    <database>Unknown</database>
  </source>
  <size>
    <width>704</width>
    <height>480</height>
    <depth>3</depth>
  </size>
  <segmented>0</segmented>
  <object>
    <name>pedestrianCrossing</name>
    <pose>Unspecified</pose>
    <truncated>0</truncated>
    <difficult>0</difficult>
    <bndbox>
      <xmin>539</xmin>
      <ymin>254</ymin>
      <xmax>502</xmax>
      <ymax>287</ymax>
    </bndbox>
  </object>
</annotation>

**Comment by 1NNcoder** [comment]
`<?xml version="1.0" ?>
<annotation verified="yes">
  <folder>/home/ubuntu/datasets/img/</folder>
  <filename>pedestrian_1323820836.avi_image13.jpg</filename>
  <path>/home/ubuntu/datasets/img/pedestrian_1323820836.avi_image13.jpg</path>
  <source>
    <database>Unknown</database>
  </source>
  <size>
    <width>704</width>
    <height>480</height>
    <depth>3</depth>
  </size>
  <segmented>0</segmented>
  <object>
    <name>pedestrianCrossing</name>
    <pose>Unspecified</pose>
    <truncated>0</truncated>
    <difficult>0</difficult>
    <bndbox>
      <xmin>539</xmin>
      <ymin>254</ymin>
      <xmax>502</xmax>
      <ymax>287</ymax>
    </bndbox>
  </object>
</annotation>`

**Comment by 1NNcoder** [comment]
@thtrieu  here is the xml file, please let me know
https://gist.github.com/1NNcoder/21492c6c5b2718d4225796139464651f

**Comment by 1NNcoder** [comment]
I also tried with this xmin xmax format https://gist.github.com/1NNcoder/468b6ccef3e3e8357ba92d30e627697a

**Comment by thtrieu** [comment]
Well, from what I see in the xml file, 
![image](https://cloud.githubusercontent.com/assets/9194247/24846542/7e8f3b66-1de5-11e7-932a-ecf39276f395.png)
xmin is clearly > xmax, which is the annotation's fault.

**Comment by 1NNcoder** [comment]
I also tried it the correct way 
https://www.evernote.com/l/ACWQe2zDRx5A_7onIic16HBIHMwCHjYBR5g

**Comment by thtrieu** [comment]
Still the same error for the same xml file?

**Comment by 1NNcoder** [comment]
Yes it gives same warning and loss is nan

**Comment by thtrieu** [comment]
Try deleting `net/yolo/parse-history.txt` and parse again.

**Comment by thtrieu** [comment]
Is everything okay now? If things work, please report here for other people in the future & close the issue.

**Comment by 1NNcoder** [comment]
I will try it out in 12 hours from now when I go back to the office.

**Comment by 1NNcoder** [comment]
yes that fixed it. Thank you so much.
So just as a recap for others,
my bug was that my xmin was greater than xmax
once I fixed that and did `rm net/yolo/parse-history.txt` 
I was working

**Comment by LilNader** [comment]
Hello, I have the same problem and I cannot find that parse-history.txt file in net/yolo/ directory to delete it.
I use windows 10

**Comment by MukundaM** [comment]
I was also facing the same issue.

Later I checked my xmin, xmax, ymin, and ymax coordinates in xml annotations files in which some of the coordinates were wrong.
Like, _xmin was > xmax or ymin was > ymax which is absolutely wrong._

After removing such errors my system worked fine and is working very well now.
Before starting training check whether your annotations have some files with such errors.

Thank You.
